structures/segment-tree/segment-tree.tex

```diff
@@ -0,0 +1,99 @@
+For a given array of size $n$, a segment tree constructed from it will use up to $2^{\lceil \log_2 (n)\rceil + 1} - 1$ space.
+Once built, the structure of a segment tree cannot change.
+
+% #ifdef hackpack
+
+\acmlisting[label=Segment Tree Reference Code, caption=Segment Tree Reference Code]{./structures/segment-tree/segment-tree.cpp}
+
+% #endif
+
+
+% #ifdef hackpackpp
+
+Generally, a segment tree supports two operations: update and query.
+The update operation will account for changes to values in the source dataset.
+It is even possible to implement a range update operation to change an entire range of values.
+The query operation allows for fast retrieval about segment information.
+In the case of the prefix sum, the update operation will modify the tree's nodes such that it will contain correct information after changes, and the the query operation can return a sum for values in an arbitrary range.
+
+\subsection{Applications}
+
+\begin{itemize}
+    \item range sums in a frequently changing dataset
+    \item range minimum/maximum queries
+\end{itemize}
+
+\subsection{Example Contest Problem: Coming and Going}
+
+Farmer John is considering expanding the cows' barn, but he would first like to learn just how many of his cows spend any given time of day in the barn.
+In consideration of expanding the cows' barn, Farmer John decided to see just how many of his cows spent any given time of day in the barn.
+To do so, he installed motion sensors on each entryway/exit to the building.
+These sensors are able to detect both the entry and exit of a warm-blooded being.
+This, coupled with his cows' tags, allows him to monitor the movements of specific cows.
+
+The night after the full day in which the sensors were installed, Farmer John sat down to do what he thought was some straightforward analysis, but quickly realized that he was in over his head with only mental math.
+He dusts off his old Commodore 64 and begins tabulating the data that he has collected.
+Among this data are the times that individual cows were inside the barn.
+\textit{While} he enters these times into the computer for each cow, Farmer John might sometimes get curious about how many cows were inside the barn during a specific time range.
+
+Help him figure out how many cows were inside the barn during a specific time range as he (slowly) tabulates the data.
+
+\subsubsection{Input Format}
+
+\begin{itemize}
+    \item Line 1: A single integer, $C$, representing the number of cows Farmer John has data on.
+
+    \item The following lines are placed in $C$ groups detailed as follows:
+    \begin{itemize}
+        \item A single line containing two integers, $t_1$ and $t_2$ representing a time range for which Farmer John wants a count of cows present.
+        These values are guaranteed to fall between 6 and 20 (inclusive), representing the times between and including 6AM to 8PM, except for the case when both are 0.
```

Comment:
Alternatively, the `0 0` case doesn't add to much to the problem algorithmically and complicates the problem.  Perhaps we could just remove it.